lapce-ui/src/tab.rs

```diff
@@ -945,27 +946,76 @@ impl LapceTab {
                     }
-                    LapceUICommand::VoltRemoved(volt) => {
+                    LapceUICommand::VoltInstalling(volt, error) => {
                         let plugin = Arc::make_mut(&mut data.plugin);
-                        let id = volt.id();
-                        plugin.installed.remove(&id);
-                        if plugin.disabled.contains(&id) {
-                            plugin.disabled.remove(&id);
-                            let _ = data.db.save_disabled_volts(
-                                plugin.disabled.iter().collect(),
+
+                        if let Some(elem) = plugin.installing.get_mut(&volt.id()) {
+                            if !error.is_empty() {
+                                elem.set_error(error);
+                            }
+                        } else {
+                            plugin.installing.insert(
+                                volt.id(),
+                                PluginInstallStatus::new(
+                                    PluginInstallType::INSTALLATION,
+                                    &volt.display_name,
+                                    error.to_string(),
+                                ),
                             );
                         }
-                        if plugin.workspace_disabled.contains(&id) {
-                            plugin.workspace_disabled.remove(&id);
-                            let _ = data.db.save_disabled_volts(
-                                plugin.workspace_disabled.iter().collect(),
+                    }
+                    LapceUICommand::VoltRemoving(volt, error) => {
+                        let plugin = Arc::make_mut(&mut data.plugin);
+
+                        if let Some(elem) = plugin.installing.get_mut(&volt.id()) {
+                            if !error.is_empty() {
+                                elem.set_error(error);
+                            }
+                        } else {
+                            plugin.installing.insert(
+                                volt.id(),
+                                PluginInstallStatus::new(
+                                    PluginInstallType::UNINSTALLATION,
+                                    &volt.display_name,
+                                    error.to_string(),
+                                ),
                             );
                         }
                     }
+                    LapceUICommand::VoltRemoved(volt, only_installing) => {
+                        let plugin = Arc::make_mut(&mut data.plugin);
+                        let id = volt.id();
+
+                        // if there is a value inside the installing map, remove it from there as soon as it is installed.
+                        plugin.installing.remove(&volt.id());
+
+                        if !(*only_installing) {
+                            plugin.installed.remove(&id);
+
+                            if plugin.disabled.contains(&id) {
```

Comment:
You don't need the `contains` check. [remove](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.remove) will tell you if the element was present

Reply:
Sorry I might have miscommunicated. You still need the `if` to not save data unnecessarily. Like `if plugin.disabled.remove(...) {`

Reply:
Oh oops, yes of course!